cloudinit/cmd/main.py

```diff
@@ -210,6 +210,37 @@ def attempt_cmdline_url(path, network=True, cmdline=None):
             (cmdline_name, url, path))
 
 
+def purge_cache_on_python_version_change(init):
+    """Purge the cache if python version changed on us.
+
+    There could be changes not represented in our cache (obj.pkl) after we
+    upgrade to a new version of python, so at that point clear the cache
+    """
+    current_python_version = '%d.%d' % (
+        sys.version_info.major, sys.version_info.minor
+    )
+    python_version_path = os.path.join(
+        init.paths.get_cpath('data'), 'python-version'
+    )
+    if os.path.exists(python_version_path):
+        cached_python_version = open(python_version_path).read()
+        # The Python version has changed out from under us, anything that was
+        # pickled previously is likely useless due to API changes.
+        if cached_python_version != current_python_version:
+            LOG.debug('Python version change detected. Purging cache')
+            init.purge_cache(True)
+            util.write_file(python_version_path, current_python_version)
+    else:
+        if os.path.exists(init.paths.get_ipath_cur('obj_pkl')):
+            # Only worrying about purging if there's cached instance data
+            LOG.debug(
+                'Could not determine Python version used to write cache. '
+                'Purging cache'
+            )
+            init.purge_cache(True)
```

Comment:
Thanks for this PR and discussion, I'm good with the principle of purging cache across python minor version upgrades as you are more likely to stumble over inconsistencies in stdlib dependencies that cloud-init doesn't control. I'm just not thrilled with this big hammer here in the "else" clause.

What it looks like we are doing here is blowing away the cache on all systems which upgrade cloud-init, because they don't yet track the python_version against which the pickle was created. purging all caches due to a cloud-init package upgrade then exposes all systems, regardless of how old they are and whether or not they have changed their python version, to forcibly re-run datasource detection and attempt th crawl their cloud metadata services. The problem in this case is that those systems could no longer have access to their metadata or configdrives etc and cannot possibly redetect or re-run the same setup.

 Wouldn't it instead make more sense for us to just skip this whole else clause, and just call the write_file so that subsequent python major.minor upgrades get triggered?

Alternatively, distro packaging could also remedy this is within the packaging scripts of cloud-init, in the postinst or %post scripts for the package. If postinst script has visibility to the package we are upgrading from (as debian postinst does) and we know the packaging versions present on RHEL7 vs RHEL8 triggers this change, couldn't the postinst say if prev_ver <= "<VER_FROM_RHEL7>" and current_ver >= "<VER_FROM_RHEL8>" then rm obj.pkl.


I'm trying to come up with alternatives because I really feel like this bug we are trying to fix here seems datasource-specific across an even more specific python major upgrade. I don't want all cloud-init consumers on every distro and to retrigger a metadata crawl across package upgrade when this is really a corner case user-story.  If we are able to drop the 'else' I have no qualms with instrumenting this solution "going-forward". It just means that RHEL7 would need to see an upgrade to cloud-init 21.X (which persists the python version) in order to support upgrades. (Or help documentation/script that tells a customer to write out python major.minor to /var/lib/cloud/instance/data/python-version before major distro upgrade)

Reply:
@blackboxsw , yes I agree. I didn't realize that it would trigger a metadata recrawl on every single upgraded instance. I pushed a new commit with that part removed.

In the shorter term, I agree that i makes more sense for things to be handled in a distro-specific way. If a distro can handle removing the pickle in their packaging/upgrade code, that seems ideal.

Reply:
Yes no longer having access to the config-drive or metadata service is something I did not consider, thanks for catching this. As such purging blindly is not a good idea. That said the idea was to help the user which we miss again with the latest push from @TheRealFalcon and the change of the `else` clause because we just write the Python version, then at some point later the cache read may fail and we are in the same situation as we are today, meaning it is not trivial to figure out what's going on when the cache doesn't load. So at the very least, if we do not want to blindly clean the cache, which I agree with, we should write a message to the log that leaves a bread crum for debugging purposes.

Reply:
> we should write a message to the log that leaves a bread crum for debugging purposes.

@rjschwei , Can you explain what you mean by this? What would we write? When would we write it?

Is there code that could be added as part of the distro upgrade process? Once the initial upgrade has been performed to a version of cloud-init that has this change, this shouldn't be a problem anymore.

Reply:
@TheRealFalcon I am not fond of the "what the distro can do"idea in this case as we will simply get different behavior across distros and that makes for an inconsistent user experience. Which is the reason I worked on this in the first place rather than just clearing the cache in %post and calling it a day. So if we go down the "distro package should do something path" I think we would need to clearly document what we want that something to be in the hopes that all packagers will adhere to our recommendation.

Also that eliminates the "someone has a cache" and builds their own Python environment + cloud-init use case. We can argue how important that is but the use case probably exists.

As far as the breadcrumbs are concerned the issue is that at the place in the process when we create the py_ver file we do not yet know if the cache can actually be loaded. So we might write the file when the cache should be purged. In this case we could emit a message "Writing py_ver file, but cache compatibility status is unknown" or something more verbose. Then we would at least leave a hint in the log that if there is a cache loading issue that the cache should have been purged to begin with.

Reply:
>  I am not fond of the "what the distro can do"idea in this case as we will simply get different behavior across distros and that makes for an inconsistent user experience. Which is the reason I worked on this in the first place rather than just clearing the cache in %post and calling it a day.

I understand the frustration, but do we have another option? We're already agreed that clearing the cache for ALL users is too disruptive. The only other way this could be fixed upstream is to do the thing that OddBloke initially suggested and use the upgrade framework to fix the root issue here:
`On Python 3.6, MIME multipart has a new required attribute “policy” that didn’t exist in Python 2.7.`.

If we don't do that, then the solution has to be distro-specific, right? If we do fix that, then (along with this cache clear on version change) this issue should disappear entirely. I'm now leaning towards these two things together being the optimal solution here.

> In this case we could emit a message "Writing py_ver file, but cache compatibility status is unknown" or something more verbose. Then we would at least leave a hint in the log that if there is a cache loading issue that the cache should have been purged to begin with.

That makes sense. If the cache exists and the python version file doesn't exist, I think logging a one-time warning is reasonable...but we shouldn't need to do this (or a packaging-specific solution) if we fix it via the upgrade framework.

Reply:
Sorry if there is a misunderstanding. I am not advocating to revisit previous conversations. IMHO using the upgrade framework for this puts us into a more or less continuous chasing loop and thus I think that's not a good idea. I suggest that we add a logging message to the current proposal in the `else:` code branch and call it good.

Reply:
> IMHO using the upgrade framework for this puts us into a more or less continuous chasing loop and thus I think that's not a good idea.

I'm suggesting we do both. Purging the cache on python version upgrade fixes the issue so we don't have a continuous chasing loop, but it still doesn't fix this MIME issue that happens when upgrading to the fixed version. Fixing that with the upgrade framework can be a "one off" fix for a problem that we'll no longer see with the python version checking in place.

If we do both of these things together, then there's no need for distro-specific upgrade code. I think it's the easiest path forward.

We can also add a logging message.

Reply:
@rjschwei and @blackboxsw , does this plan work for you? If so, I think we can merge this current PR as-is, then I can create a new PR to fix the specific MIME issue that was mentioned here.

Reply:
@TheRealFalcon LGTM Thanks